api/src/main/java/io/minio/MinioClient.java

```diff
@@ -932,43 +1031,49 @@ private void updateRegionCache(String bucketName)
            InternalException {
     if (bucketName != null && this.accessKey != null && this.secretKey != null
         && !BucketRegionCache.INSTANCE.exists(bucketName)) {
-      Map<String,String> queryParamMap = new HashMap<>();
-      queryParamMap.put("location", null);
-
-      HttpResponse response = execute(Method.GET, US_EAST_1, bucketName, null,
-                                      null, queryParamMap, null, 0);
-
-      // existing XmlEntity does not work, so fallback to regular parsing.
-      XmlPullParser xpp = xmlPullParserFactory.newPullParser();
-      String location = null;
-
-      xpp.setInput(response.body().charStream());
-      while (xpp.getEventType() != xpp.END_DOCUMENT) {
-        if (xpp.getEventType() ==  xpp.START_TAG && xpp.getName() == "LocationConstraint") {
+      // Compute bucket region but always pick this.region if the latter is set,
+      // otherwise, call the bucket location API.
+      String region;
+      if (this.region != null && !this.region.equals("")) {
+        region = this.region;
```

Comment:
This code looks incorrect - bucketCache is still used when the region is set. we should refactor it so that when region is given we never ever use the bucket cache instead the return value is directly from `this.region` 

Region cache should only be used when `this.region` is not set - this would indicate the future code is not added on top to have a mutating state underneath.